add query settings interface

Yancey0623 opened the pull request:
fixed #45

Yancey0623 committed:
add query settings interface

Yancey0623 committed:
Merge remote-tracking branch 'origin' into sql_settings

Yancey0623 committed:
fix ci

weiguoz commented on driver.go:
```diff
@@ -8,16 +8,26 @@ import (
 // register driver
 func init() {
-	sql.Register("maxcompute", &Driver{})
+	sql.Register("maxcompute", &ODPSDriver{})
```
I'd rather use `MaxcomputeDriver` than `ODPSDriver`. 
In fact, we use `odps*` inside the repo but do not expose it.

Yancey0623 replied:
Right, maxcompute is the open-source name.

Yancey0623 committed:
update by comment

weiguoz commented on driver.go:
```diff
@@ -8,16 +8,26 @@ import (
-type Driver struct{}
+// ODPSDriver impls database/sql/driver.Driver
+type ODPSDriver struct {
+	conn odpsConn
```
I'm not sure if it is reasonable to maintain a `conn` inside the driver.

Yancey0623 replied:
Maybe it's okay, we need to maintain the `settings` struct in the DB object and the `Driver` struct can be exposed when we implement the `sql.database`.

tonyyang-svail commented on driver.go:
```diff
@@ -8,16 +8,26 @@ import (
+// SetQuerySettings sets the global query settings.
+// TODO(Yancey1989): add the one-off query settings interface.
+func (o MaxcomputeDriver) SetQuerySettings(hints map[string]string) error {
+	o.conn.hints = hints
```
deep copy or shallow copy?

tonyyang-svail reviewed (commented):
@Yancey1989 @weiguoz Please be aware that after this change, `odpsConn` is no longer reentrant. Multiple goroutines can't share the same `odpsConn`. For example,

```go
db := sql.Open("maxcompute", ...)

go func() {
    db.Driver().(*MaxcomputeDriver).SetQuerySettings(...)
}()

go func() {
    db.Query(...) // Oops, the setting has been changed. Or even worse, panic due to the race on hints.
}()
```

I like the idea `// TODO(Yancey1989): add the one-off query settings interface.`. I am wondering if the following implementation is possible.

```go
package maxcompute

func QueryWithHint(db sql.DB, query string, hints map[string]string) (driver.Rows, error)  {
    conn, ok := db.Driver().(*MaxcomputeDriver)
    if !ok {
        return nil, fmt.Errorf(...)
    }
    newConn := odpsConn{conn.Client, conn.cfg, hints}
    return newConn.Query(query, ...)
}
```

Then in another package, for example

```go
package main

func main() {
    db := Open("maxcompute", ...)
    query := ...
    hint := ...
    rows, err := maxcompute.QueryWithHint(conn, query, hints)
}

Yancey0623 commented:
Thanks for @tonyyang-svail 's reminding, in the `pyodps`, there are two ways to set the query settings, the global one:

``` python
>>> from odps import options
>>> options.sql.settings = {'odps.sql.mapper.split.size': 16}
>>> o.execute_sql('select * from pyodps_iris')  # 会根据全局配置添加hints
```

and the one-off one:
``` python
o.execute_sql('select * from pyodps_iris', hints={'odps.sql.mapper.split.size': 16})
```
ref: https://pyodps.readthedocs.io/zh_CN/latest/base-sql.html#sql-hints

Maybe `gomaxcompute` need to implement both the two APIs. I think `QueryWithHint ` is a good way to implement the one-off API, will implement it in this PR.

tonyyang-svail commented on driver_test.go:
```diff
@@ -21,3 +21,12 @@ func TestSQLOpen(t *testing.T) {
+	a := assert.New(t)
+	db, err := sql.Open("maxcompute", cfg4test.FormatDSN())
+	a.NoError(err)
+	db.Driver().(*MaxcomputeDriver).SetQuerySettings(map[string]string{"odps.sql.mapper.split.size": "16"})
```
Does this test verify the hints is effective?

tonyyang-svail replied:
Shall we also test `odps.sql.mode = script`?

Yancey0623 replied:
> Does this test verify the hints is effective?

No, it's not easy to test `odps.sql.mapper.split.size ` on the client-side. But maybe test UDF script is good way, will add this one.

Yancey0623 commented:
The script mode of Maxcompute on Aliyun: https://help.aliyun.com/document_detail/90082.html?spm=a2c4g.11186623.6.692.5036612drr5apU

tonyyang-svail commented:
@Yancey1989 btw, do we need to support logging on a job?

Yancey0623 commented:
> @Yancey1989 btw, do we need to support logging on a job?

@tonyyang-svail , I think we need that, the logs message is useful such as the job view link.

weiguoz reviewed (dismissed):
LGTM

Yancey0623 commented:
Hi @tonyyang-svail , I file an issue to talk about the one-off API #50, maybe we can continue to talk about the API design there.

Yancey0623 committed:
query hints in dsn

Yancey0623 commented:
Hi @weiguoz @tonyyang-svail @typhoonzero 
I update this PR that define the query hints arguments in DSN:
``` text
access_id:access_key@service.com/api?curr_project=test_ci&scheme=http&hints_odps.sql.mapper.split_size=16
```
the URL query arguments key has prefix `hints_` would be converted to query hints argument.

typhoonzero commented on dsn.go:
```diff
@@ -26,32 +28,41 @@ func ParseDSN(dsn string) (*Config, error) {
+	for k, v := range querys {
+		// The query args such as hints_odps.sql.mapper.split_size=16
+		// would be converted to the maxcompute query hints: {"odps.sql.mapper.split_size": "16"}
+		if strings.HasPrefix(k, "hints_") {
```
Can we use `odps.` as prefix?

Yancey0623 replied:
Maybe we can't, some case like UDF prediction job would set the query hints using `mst` as the prefix:

``` sql
set mst.model.path=oss://<oss path>;
set mst.model.name=<model name>
```

weiguoz commented on dsn.go:
```diff
@@ -26,32 +28,41 @@ func ParseDSN(dsn string) (*Config, error) {
+	for k, v := range querys {
+		// The query args such as hints_odps.sql.mapper.split_size=16
+		// would be converted to the maxcompute query hints: {"odps.sql.mapper.split_size": "16"}
+		if strings.HasPrefix(k, "hints_") {
```
How about using `hint` as a prefix?
Because the `DSN` maybe looks like `gomaxcompute://user:passwd@host/project?hint_odps.sql.mapper.split_size=16&hint_odps.sql.other.key=value`, hints separated by a `&`.

Yancey0623 replied:
You're right, thanks and will fix that.

Yancey0623 committed:
hints to hint

Yancey0623 committed:
fix unit test

Yancey0623 committed:
fix ut

typhoonzero commented on dsn.go:
```diff
@@ -26,32 +30,41 @@ func ParseDSN(dsn string) (*Config, error) {
-	if query["scheme"] != "http" && query["scheme"] != "https" {
+
+	for k, v := range querys {
+		// The query args such as hints_odps.sql.mapper.split_size=16
```
hints_ => hint?

Yancey0623 replied:
fixed

typhoonzero reviewed (dismissed):
LGTM

typhoonzero commented:
Is this ready to merge? @Yancey1989

Yancey0623 committed:
fix comment
